### ISSUE TITLE
segfault in runtest.py

### ISSUE THREAD

**Issue by robertnishihara** [amplab/ray#170]
Sometimes the Travis segfaults in `runtest.py` on Mac OS X with the message
```
$ python runtest.py
scheduler: writing to log file /tmp/raylogs/2016-06-26=20:17:34-scheduler.log
D0626 20:17:34.866177000 140735276749568 ev_posix.c:101] Using polling engine: poll
object store: writing to log file /tmp/raylogs/2016-06-26=20:17:35-objstore-127.0.0.1:20001.log
D0626 20:17:35.026280000 140735276749568 ev_posix.c:101] Using polling engine: poll
D0626 20:17:35.576193000 140735276749568 ev_posix.c:101] Using polling engine: poll
D0626 20:17:36.901697000 140735276749568 ev_posix.c:101] Using polling engine: poll
Attempting to kill process at address 127.0.0.1:10001.
Successfully killed process at address 127.0.0.1:10001.
Attempting to kill process at address 127.0.0.1:20001.
Successfully killed process at address 127.0.0.1:20001.
Attempting to kill process at address 127.0.0.1:40001.
Successfully killed process at address 127.0.0.1:40001.
.scheduler: writing to log file /tmp/raylogs/2016-06-26=20:17:37-scheduler.log
D0626 20:17:37.347365000 140735276749568 ev_posix.c:101] Using polling engine: poll
object store: writing to log file /tmp/raylogs/2016-06-26=20:17:37-objstore-127.0.0.1:20002.log
D0626 20:17:37.460378000 140735276749568 ev_posix.c:101] Using polling engine: poll
D0626 20:17:39.912683000 140735276749568 ev_posix.c:101] Using polling engine: poll
D0626 20:17:40.093953000 140735276749568 ev_posix.c:101] Using polling engine: poll
D0626 20:17:40.122660000 140735276749568 ev_posix.c:101] Using polling engine: poll
Attempting to kill process at address 127.0.0.1:10002.
E0626 20:17:40.251382000 140735276749568 tcp_client_posix.c:173] failed to connect to 'ipv4:127.0.0.1:10002': socket error: connection refused
Successfully killed process at address 127.0.0.1:10002.
Attempting to kill process at address 127.0.0.1:20002.
E0626 20:17:40.311871000 140735276749568 tcp_client_posix.c:173] failed to connect to 'ipv4:127.0.0.1:10002': socket error: connection refused
Successfully killed process at address 127.0.0.1:20002.
Attempting to kill process at address 127.0.0.1:40003.
Successfully killed process at address 127.0.0.1:40003.
Attempting to kill process at address 127.0.0.1:40004.
Successfully killed process at address 127.0.0.1:40004.
Attempting to kill process at address 127.0.0.1:40005.
Successfully killed process at address 127.0.0.1:40005.
.scheduler: writing to log file /tmp/raylogs/2016-06-26=20:17:40-scheduler.log
D0626 20:17:40.496432000 140735276749568 ev_posix.c:101] Using polling engine: poll
object store: writing to log file /tmp/raylogs/2016-06-26=20:17:40-objstore-127.0.0.1:20003.log
D0626 20:17:40.609605000 140735276749568 ev_posix.c:101] Using polling engine: poll
D0626 20:17:41.120116000 140735276749568 ev_posix.c:101] Using polling engine: poll
D0626 20:17:42.252428000 140735276749568 ev_posix.c:101] Using polling engine: poll
Attempting to kill process at address 127.0.0.1:10003.
Successfully killed process at address 127.0.0.1:10003.
Attempting to kill process at address 127.0.0.1:20003.
Successfully killed process at address 127.0.0.1:20003.
Attempting to kill process at address 127.0.0.1:40007.
Successfully killed process at address 127.0.0.1:40007.
.scheduler: writing to log file /tmp/raylogs/2016-06-26=20:17:42-scheduler.log
D0626 20:17:42.844424000 140735276749568 ev_posix.c:101] Using polling engine: poll
object store: writing to log file /tmp/raylogs/2016-06-26=20:17:42-objstore-127.0.0.1:20004.log
D0626 20:17:42.950088000 140735276749568 ev_posix.c:101] Using polling engine: poll
object store: writing to log file /tmp/raylogs/2016-06-26=20:17:43-objstore-127.0.0.1:20005.log
D0626 20:17:43.465256000 140735276749568 ev_posix.c:101] Using polling engine: poll
Attempting to kill process at address 127.0.0.1:10004.
Successfully killed process at address 127.0.0.1:10004.
Attempting to kill process at address 127.0.0.1:20004.
Successfully killed process at address 127.0.0.1:20004.
Attempting to kill process at address 127.0.0.1:20005.
Successfully killed process at address 127.0.0.1:20005.
/Users/travis/build.sh: line 45: 19498 Segmentation fault: 11  python runtest.py
The command "python runtest.py" exited with 139.
```

I can reproduce this error on Mac OS X (not deterministically) by removing everything from `runtest.py` except for `ObjStoreTest`. Running the test in gdb seems to indicate that the error occurs when an object reference goes out of scope and its destructor calls `decrement_reference_count`, which in turn calls `scheduler_stub_->DecrementRefCount(&context, request, &reply)`, which segfaults.

The gdb backtrace is
```
Program received signal SIGSEGV, Segmentation fault.
0x00000001011537ce in Scheduler::Stub::DecrementRefCount(grpc::ClientContext*, DecrementRefCountRequest const&, AckReply*) () from /Users/fortesting/ray/lib/python/ray/libraylib.so
#0  0x00000001011537ce in Scheduler::Stub::DecrementRefCount(grpc::ClientContext*, DecrementRefCountRequest const&, AckReply*) () from /Users/fortesting/ray/lib/python/ray/libraylib.so
#1  0x00000001010c2c5b in Worker::decrement_reference_count(std::__1::vector<unsigned long, std::__1::allocator<unsigned long> >&) () from /Users/fortesting/ray/lib/python/ray/libraylib.so
#2  0x00000001010aff69 in PyObjRef_dealloc(PyObjRef*) () from /Users/fortesting/ray/lib/python/ray/libraylib.so
#3  0x0000000100044333 in list_dealloc () from /anaconda/lib/libpython2.7.dylib
#4  0x000000010003cd8d in frame_dealloc () from /anaconda/lib/libpython2.7.dylib
#5  0x00000001000c57e3 in PyEval_EvalFrameEx () from /anaconda/lib/libpython2.7.dylib
#6  0x00000001000c83a3 in PyEval_EvalCodeEx () from /anaconda/lib/libpython2.7.dylib
#7  0x000000010003e100 in function_call () from /anaconda/lib/libpython2.7.dylib
#8  0x000000010000c552 in PyObject_Call () from /anaconda/lib/libpython2.7.dylib
#9  0x00000001000c2538 in PyEval_EvalFrameEx () from /anaconda/lib/libpython2.7.dylib
#10 0x00000001000c83a3 in PyEval_EvalCodeEx () from /anaconda/lib/libpython2.7.dylib
#11 0x000000010003e100 in function_call () from /anaconda/lib/libpython2.7.dylib
#12 0x000000010000c552 in PyObject_Call () from /anaconda/lib/libpython2.7.dylib
#13 0x000000010001dbed in instancemethod_call () from /anaconda/lib/libpython2.7.dylib
#14 0x000000010000c552 in PyObject_Call () from /anaconda/lib/libpython2.7.dylib
#15 0x00000001000796aa in slot_tp_call () from /anaconda/lib/libpython2.7.dylib
#16 0x000000010000c552 in PyObject_Call () from /anaconda/lib/libpython2.7.dylib
#17 0x00000001000c0013 in PyEval_EvalFrameEx () from /anaconda/lib/libpython2.7.dylib
#18 0x00000001000c83a3 in PyEval_EvalCodeEx () from /anaconda/lib/libpython2.7.dylib
#19 0x000000010003e100 in function_call () from /anaconda/lib/libpython2.7.dylib
#20 0x000000010000c552 in PyObject_Call () from /anaconda/lib/libpython2.7.dylib
#21 0x00000001000c2538 in PyEval_EvalFrameEx () from /anaconda/lib/libpython2.7.dylib
#22 0x00000001000c83a3 in PyEval_EvalCodeEx () from /anaconda/lib/libpython2.7.dylib
#23 0x000000010003e100 in function_call () from /anaconda/lib/libpython2.7.dylib
#24 0x000000010000c552 in PyObject_Call () from /anaconda/lib/libpython2.7.dylib
#25 0x000000010001dbed in instancemethod_call () from /anaconda/lib/libpython2.7.dylib
#26 0x000000010000c552 in PyObject_Call () from /anaconda/lib/libpython2.7.dylib
#27 0x00000001000796aa in slot_tp_call () from /anaconda/lib/libpython2.7.dylib
#28 0x000000010000c552 in PyObject_Call () from /anaconda/lib/libpython2.7.dylib
#29 0x00000001000c0013 in PyEval_EvalFrameEx () from /anaconda/lib/libpython2.7.dylib
#30 0x00000001000c83a3 in PyEval_EvalCodeEx () from /anaconda/lib/libpython2.7.dylib
#31 0x000000010003e100 in function_call () from /anaconda/lib/libpython2.7.dylib
#32 0x000000010000c552 in PyObject_Call () from /anaconda/lib/libpython2.7.dylib
#33 0x00000001000c2538 in PyEval_EvalFrameEx () from /anaconda/lib/libpython2.7.dylib
#34 0x00000001000c83a3 in PyEval_EvalCodeEx () from /anaconda/lib/libpython2.7.dylib
#35 0x000000010003e100 in function_call () from /anaconda/lib/libpython2.7.dylib
#36 0x000000010000c552 in PyObject_Call () from /anaconda/lib/libpython2.7.dylib
#37 0x000000010001dbed in instancemethod_call () from /anaconda/lib/libpython2.7.dylib
#38 0x000000010000c552 in PyObject_Call () from /anaconda/lib/libpython2.7.dylib
#39 0x00000001000796aa in slot_tp_call () from /anaconda/lib/libpython2.7.dylib
#40 0x000000010000c552 in PyObject_Call () from /anaconda/lib/libpython2.7.dylib
#41 0x00000001000c0013 in PyEval_EvalFrameEx () from /anaconda/lib/libpython2.7.dylib
#42 0x00000001000c57bf in PyEval_EvalFrameEx () from /anaconda/lib/libpython2.7.dylib
#43 0x00000001000c57bf in PyEval_EvalFrameEx () from /anaconda/lib/libpython2.7.dylib
#44 0x00000001000c83a3 in PyEval_EvalCodeEx () from /anaconda/lib/libpython2.7.dylib
#45 0x000000010003e100 in function_call () from /anaconda/lib/libpython2.7.dylib
#46 0x000000010000c552 in PyObject_Call () from /anaconda/lib/libpython2.7.dylib
#47 0x000000010001dbed in instancemethod_call () from /anaconda/lib/libpython2.7.dylib
#48 0x000000010000c552 in PyObject_Call () from /anaconda/lib/libpython2.7.dylib
#49 0x00000001000792d8 in slot_tp_init () from /anaconda/lib/libpython2.7.dylib
#50 0x0000000100076155 in type_call () from /anaconda/lib/libpython2.7.dylib
#51 0x000000010000c552 in PyObject_Call () from /anaconda/lib/libpython2.7.dylib
#52 0x00000001000c0013 in PyEval_EvalFrameEx () from /anaconda/lib/libpython2.7.dylib
#53 0x00000001000c83a3 in PyEval_EvalCodeEx () from /anaconda/lib/libpython2.7.dylib
#54 0x00000001000c84c6 in PyEval_EvalCode () from /anaconda/lib/libpython2.7.dylib
#55 0x00000001000ede5e in PyRun_FileExFlags () from /anaconda/lib/libpython2.7.dylib
#56 0x00000001000ee0fa in PyRun_SimpleFileExFlags () from /anaconda/lib/libpython2.7.dylib
#57 0x000000010010517d in Py_Main () from /anaconda/lib/libpython2.7.dylib
#58 0x0000000100000f54 in start ()
warning: (Internal error: pc 0x0 in read in psymtab, but not in symtab.)
```

**Comment by robertnishihara** [comment]
Addressed by #212. See that PR for an explanation of the problem.

**Comment by robertnishihara** [comment]
This came up again on Mac OS X (reported by @naltieri). #237 should address it.
